grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,108 @@
+package org.pih.warehouse.invoice
+
+import grails.gorm.transactions.Transactional
+
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.order.OrderAdjustment
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.order.OrderItemStatusCode
+import org.pih.warehouse.shipping.ShipmentItem
+
+/**
+ * This service exists for the sole purpose of performing data migrations for changelog 0.9.x/2024-08-20-0000.
+ * The logic was only put in a service class so that it could be unit tested. Do not use it in regular flows.
+ */
+@Transactional
+class PrepaymentInvoiceMigrationService {
+
+    PrepaymentInvoiceService prepaymentInvoiceService
+
+    /**
+     * For STEP 1 of the migration.
+     *
+     * Set the amount field for all prepayment invoice items that existed before the partial invoicing
+     * feature was introduced.
+     */
+    void updateAmountFieldForPrepaymentInvoiceItems() {
+        InvoiceType prepaymentInvoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+
+        Invoice.findAllByInvoiceType(prepaymentInvoiceType).each { Invoice prepaymentInvoice ->
+            // The amount field will be null for all pre-existing invoices since the field was not being used.
+            prepaymentInvoice.invoiceItems.find { it.amount == null }.each { InvoiceItem prepaymentInvoiceItem ->
+                prepaymentInvoiceItem.amount = computePrepaymentInvoiceItemAmount(prepaymentInvoiceItem)
+                prepaymentInvoiceItem.save(failOnError: true, flush: true)
+            }
+        }
+    }
+
+    // TODO: Can we merge this with PrepaymentInvoiceService? That way if the amount calculation ever changes, it will get picked up automatically here.
+    private BigDecimal computePrepaymentInvoiceItemAmount(InvoiceItem prepaymentInvoiceItem) {
+        BigDecimal prepaymentPercent = (prepaymentInvoiceItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
```

Comment:
In my currently pending I moved this to transient on order (you could rebase it once its merged)